bip-0200.mediawiki

```diff
@@ -0,0 +1,146 @@
+<pre>
+  BIP: ???
+  Layer: Applications
+  Title: Hashed Time-Locked Collateral Contract
+  Author: Matthew Black <matthew@atomicloans.io>
+          Tony Cai <tony@atomicloans.io>
+          Atomic Loans team <info@atomicloans.io>
+  Comments-Summary: No comments yet.
+  Comments-URI: TBD
+  Status: Proposal
+  Type: Informational
+  Created: 2019-03-19
+  License: BSD-3-Clause
+           CC0-1.0
+</pre>
+
+==Abstract==
+
+This BIP describes a script for generalized debt agreement contract based on Hashed Time-Lock Contract (BIP 199) transactions according to the Atomic Loans specification (https://arxiv.org/pdf/1901.05117.pdf). For more details visit https://atomicloans.io.
+
+==Summary==
+
+A Hashed Time-Locked Collateral Contract (HTLCC) consists of two scripts that permit a designated party (the "borrower") to lock funds on the Bitcoin chain for a specified amount of time as collateral in a debt agreement where the loan principal is denominated in a currency on another blockchain (eg. Ethereum). We denote the blockchain on which the loan principal is issued the principal blockchain.
+
+The purpose of each script is to enable the creation of a debt agreement between two parties (the "borrower" and the "lender"), where the collateral is locked in a P2SH, and can only be spent once the borrower repays the principal and interest in the debt agreement on the principal blockchain. In the case that the borrower does not repay, the borrower or lender can opt for liquidation of the collateral, which will involve the atomic swapping of collateral for the loan currency. In the case that at least one of the two parties don't opt for liquidation, then each party will be entitled to a percentage of the collateral, decided when the funds are initially locked in the P2SH.
+
+These funds are locked into two scripts. Refundable Collateral and Seizable Collateral scripts. The funds sent to these scripts represent the percentage of collateral that each party is entitled to in the case that repayment fails, and the parties don't opt for liquidation.
+
+The Refundable Collateral script takes the following form:
+
+    OP_IF
+        OP_SIZE <secret b2 length> OP_EQUALVERIFY [HASHOP] <secret hash b2> OP_EQUALVERIFY OP_DUP OP_HASH160 <borrower pubkey hash> OP_EQUALVERIFY OP_CHECKSIG
```

Comment:
Why are you checking the length and the hash separately? If the hash is secure, only two identical items of the same length will match. Additionally, SHA2 hashes explicitly commit to the length.

Reply:
This script checks the length because it's being used alongside the HTLPC (initially Ethereum with ERC 1850 https://github.com/ethereum/EIPs/pull/1850). The SHA256 function in Ethereum takes up to 32 bytes, no more. Therefore, it is necessary to ensure that the secret b2 cannot be more than 32 bytes, which would allow for the secret to correspond to the correct secrethash on Bitcoin, but not Ethereum.

Reply:
Ex. `btcdeb 216443b923b04a21839c44c8f5e54c360af7b5a4541f11511584b6e413a358aa8080a8`  will be a difference result than `btcdeb 206443b923b04a21839c44c8f5e54c360af7b5a4541f11511584b6e413a358aa80a8`

Reply:
You should write this up in the (currently non-existent) Rationale section. ;)

(There's also too many Ethereum references elsewhere - remember BIPs are about Bitcoin, so probably it *only* makes sense to reference Ethereum-side stuff in Rationale)

Reply:
Great idea! Reduced talk of Ethereum and added Rationale.